### PR TITLE
Configured dynamic manifest #730

### LINKED ISSUE #730: Configure Dynamic Manifest
Dynamic Manifest support will improve PWA functionality for the site in all supported languages.
* https://www.w3.org/TR/appmanifest/

### PR DESCRIPTION
**Pull request by johnmgrant**
Closes #730

**Merge blocked until approved on dev.**

* Modified Manifest to support localization
* Updated PWA to support offline experience
* Update device icon support

(cherry picked from commit 3afc3308e8093cc5f8822cb2de30bf294ad65817)

### CODE REVIEW

**Comment by johnmgrant** [comment]
Confirmed that PWA is saved using the `start_url` of the locale that the user is set to in browser now. Will look into supporting user preferred locale setting in a separate issue.

File: messages/ui-ja.json [hunk 1]
```diff
@@ -68,7 +68,12 @@
   "Metadata": {
     "Site": {
       "Name": "ポケスリ攻略",
-      "Description": "ポケスリ攻略では、図鑑・レシピ・食材・きのみなどの様々なデータや、\nお手伝いチームの分析・ボックス内ポケモン分析といった便利なツールを提供しています。"
+      "Description": "ポケスリ攻略では、図鑑・レシピ・食材・きのみなどの様々なデータや、\nお手伝いチームの分析・ボックス内ポケモン分析といった便利なツールを提供しています。",
+      "Keywords": {
+        "PokemonSleep": "ポケモンの睡眠",
```
Comment (RaenonX) [inline messages/ui-ja.json:73]: ポケスリ should be better

File: messages/ui-zh.json [hunk 2]
```diff
@@ -68,7 +68,13 @@
   "Metadata": {
     "Site": {
       "Name": "Pokémon Sleep 攻略站",
-      "Description": "Pokémon Sleep 攻略站提供 Pokemon Sleep 中的圖鑑、食譜、食材、樹果相關資訊。同時，本站也提供如隊伍分析、寶可夢盒...等工具。"
+      "Description": "Pokémon Sleep 攻略站提供 Pokémon Sleep 中的圖鑑、食譜、食材、樹果相關資訊。同時，本站也提供如隊伍分析、寶可夢盒...等工具。",
+      "Keywords": {
+        "PokemonSleep": "Pokemon Sleep",
+        "Pokemon": "Pokemon",
+        "Pokémon": "Pokémon",
```
Comment (RaenonX) [inline messages/ui-zh.json:75]: Think this line should be deleted

File: public/icons/icon-192x192.png
Comment (RaenonX) [inline public/icons/icon-192x192.png:1]: Is the additional 4 px at the bottom intended?

File: public/icons/icon-512x512.png
Comment (RaenonX) [inline public/icons/icon-512x512.png:1]: Is the additional 4 px at the bottom intended?

File: src/app/manifest.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+
+import {generateSiteManifest} from '@/utils/manifest';
+
+
+const manifest = () => generateSiteManifest({});
```
Comment (RaenonX) [inline src/app/manifest.ts:5]: Maybe just make it a no-param function i.e. `const manifest = generateSiteManifest;` if `generateSiteManifest()` is not used elsewhere?

File: src/utils/metadata.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+          name: t('Pokedex.Main.Index.Title'),
+          url: '/pokedex',
+        },
+        {
+          name: t('About.Title'),
+          url: '/about',
+        },
+      ],
+      languages: locales.reduce((obj, _locale) => ({
+        ...obj, [_locale]: `/${_locale}`,
```
Comment (RaenonX) [inline src/utils/metadata.ts:57]: Put a linebreak here so it should be 
```typescript
{
  ...obj,
  [_locale]: `/${_locale}`,
}
```

**Review by RaenonX — CHANGES_REQUESTED** [review]

File: messages/ui-ja.json (hunk 1 above)
Comment (johnmgrant) [reply messages/ui-ja.json:73]: Just to make sure, `"PokemonSleep": "ポケスリ",`?

Comment (RaenonX) [reply messages/ui-ja.json:73]: Yeah, for `Keywords.PokemonSleep` as they usually call the game `ポケスリ` (abbreviated)

File: src/app/manifest.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+
+import {generateSiteManifest} from '@/utils/manifest';
+
+
+const manifest = () => generateSiteManifest();
```
Comment (RaenonX) [inline src/app/manifest.ts:5]: Is there any typing applicable to `manifest`? I would make it just `manifest = generateSiteManifest` if the typing kinda indicates that to save some characters here.

**Review by RaenonX — CHANGES_REQUESTED** [review]

File: src/app/manifest.ts (hunk 5 above)
Comment (johnmgrant) [reply src/app/manifest.ts:5]: Yup, can do. https://nextjs.org/docs/app/api-reference/file-conventions/metadata/manifest#generate-a-manifest-file

Comment (RaenonX) [reply src/app/manifest.ts:5]: Then type that `manifest` and just make it a `const` which is that `generateSiteManifest` function then

**Review by RaenonX — APPROVED** [review]
Good night!

**Pull request merged**